doc/making-a-visualizer.md

```diff
@@ -1,17 +1,195 @@
+from frame to frame.
+
+If you implement `setup()`, start by calling `super.setup()`, which includes
+the [`createCanvas()`](https://p5js.org/reference/#/p5/createCanvas) call that
+must appear in every p5 setup function.
+
+#### :gear: Show or stop the visualization; depart from a page element _(advanced)_
+
+You shouldn't need to implement `show()`, `stop()`, or `depart()`. You can
+learn about them from the [visualizer interface](#abstract-visualizers)
+documentation, and from how they're implemented in the `P5Visualizer` base
+class.
+
+#### :key: Draw the visualization _(required)_
+
+The `draw()` function is called on each frame, giving you a chance to draw
+your visualization! Look at the examples and the p5
+[tutorials](https://p5js.org/learn/) and
+[p5.js reference](https://p5js.org/reference/) to learn about what you can do.
+
+Drawing tools and options are found in the graphics context, `this.sketch`.
+Accessing `this.sketch` triggers some consistency checks, so it's most
+parsimonious to access it at most once per function call, storing its value as
+a local constant for future reference.
+
+Sequence information is found in `this.seq`. This is a `SequenceInterface`
+object, so it will always have a method `getElement(n)` that returns the `n`th
+element in the sequence.
+
+You have to implement this method, even if it does nothing. Your visualizer
+can't be loaded into Numberscope without it.
+
+#### :bulb: Stop and start animation _(often used)_
+
+If your visualization is a static picture, call the sketch's `noLoop()`
+function when you're done drawing. This stops the animation loop, so you don't
+waste time re-drawing the same picture dozens of times per second. If the
+visualization is only static temporarily, you can call `loop()` to re-start
+the animation loop whenever you need to.
+
+-   **p5 Visualizer Template:** Only loop while animating the white flash that
+    indicates an index change.
+-   **Differences:** Draw the whole visualization in one frame, and then call
+    `noLoop()` to stop the animation loop.
+
+#### :bulb: Respond to user interactions _(often used)_
+
+Each time the user interacts with your visualization, an
+[event handling](https://p5js.org/reference/#group-Events) function like
+`keyPressed()` or `mouseClicked()` will be called, giving you a chance to
+respond. There are handlers for a wide variety of input events.
+
+#### :key: Export the visualizer _(required)_
+
+The engine expects the visualizer to be packaged in a `VisualizerExportModule`
+object, constructed from the visualizer class and a short description string.
+
+#### :gear: Handle errors _(advanced)_
+
+You can create an error message with the `alertMessage` utility in
+`src/shared/`, and show it to the user with
+[`window.alert`](https://developer.mozilla.org/en-US/docs/Web/API/Window/alert):
+
+```typescript
+window.alert(alertMessage(someError))
+```
+
+You can also just throw an error. If it's not caught anywhere else, the
```

Comment:
Should we not add a section for documention instructions?  All visualizers should document their functionality, provide an image of the working visualizer in assets, and document all parameters in markdown inside the visualizer file, which will be automagically built into the docs, etc.

Reply:
I like that idea. I've added a sketch of where the documentation instructions could go in the visualizer-building guide.

Reply:
I was going to suggestion "as close to the top as possible," so I really like your outline. Only not resolving this because the "How to document your visualizer" section still needs to be filled in (not sure if you had any further plans for the top summary part also; it seems decent to me).